backend/capellacollab/tools/routes.py

```diff
@@ -2,31 +2,315 @@
+)
+def get_tool_by_id(tool=Depends(get_existing_tool)):
+    return tool
+
+
+@router.post(
+    "",
+    response_model=ToolTypeBase,
+    dependencies=[Depends(RoleVerification(required_role=Role.ADMIN))],
+)
+def create_tool(body: CreateTool, db: Session = Depends(get_db)):
+    return crud.create_tool(db, models.Tool(name=body.name))
+
+
+@router.put(
+    "/{tool_id}",
+    response_model=ToolTypeBase,
+    dependencies=[Depends(RoleVerification(required_role=Role.ADMIN))],
+)
+def update_tool(tool_id: int, body: CreateTool, db: Session = Depends(get_db)):
+    tool = crud.get_tool_by_id(tool_id, db)
+    return crud.update_tool(db, tool, body)
+
+
+@router.delete(
+    "/{tool_id}",
+    status_code=204,
+    dependencies=[Depends(RoleVerification(required_role=Role.ADMIN))],
+)
+def delete_tool(
+    tool: Tool = Depends(get_existing_tool), db: Session = Depends(get_db)
+):
+    if tool.id == 1:
+        raise HTTPException(
+            403,
+            {
+                "reason": "The tool 'Capella' can not be deleted.",
+            },
+        )
+    try:
+        return crud.delete_tool(db, tool)
+    except sqlalchemy.exc.IntegrityError:
+        db.rollback()
+
+        dependencies = []
+        # Search for occurrences in project-models
+        for model in projects_models_crud.get_models_by_tool(tool.id, db):
+            dependencies.append(
+                f"Model '{model.name}' in project '{model.project.name}'"
+            )
+
+        for i in range(len(dependencies) - 1):
+            dependencies[i] = dependencies[i] + ","
+
+        raise HTTPException(
+            409,
+            {
+                "reason": [
+                    f"The tool '{tool.name}' can not be deleted. Please remove the following dependencies first:"
+                ]
+                + dependencies,
+            },
+        )
+
+
+@router.get(
+    "/{tool_id}/versions",
+    response_model=list[ToolVersionBase],
+    dependencies=[Depends(RoleVerification(required_role=Role.USER))],
 )
 def get_tool_versions(tool_id: int, db: Session = Depends(get_db)):
     return crud.get_tool_versions(db, tool_id)
 
 
+@router.post(
+    "/{tool_id}/versions",
+    response_model=ToolVersionBase,
+    dependencies=[Depends(RoleVerification(required_role=Role.ADMIN))],
+)
+def create_tool_version(
+    body: CreateToolVersion,
+    tool: Tool = Depends(get_existing_tool),
+    db: Session = Depends(get_db),
+):
+    return crud.create_version(db, tool.id, body.name, False, False)
+
+
+@router.patch(
+    "/{tool_id}/versions/{version_id}",
+    response_model=ToolVersionBase,
+    dependencies=[Depends(RoleVerification(required_role=Role.ADMIN))],
+)
+def patch_tool_version(
+    body: UpdateToolVersion,
+    version=Depends(get_exisiting_tool_version),
+    db: Session = Depends(get_db),
+):
+    for key, value in body.dict().items():
+        if value is not None:
+            version.__setattr__(key, value)
```

Comment:
I still prefer my solution. This has two reasons: 
1. Your code doesn't do exactly the same. If you want to set `is_recommended` to `False`, it's not possible with your code statement. It requires the explicit `is not None`, which makes your code even longer. 
2. It is already defined in the `UpdateToolVersion` object which attributes can be updated - I think it's clear that all of these attributes get updated here as well then. 
3. When having just three attributes in the object, it might look good, but your code is getting longer and longer with every new attribute. When we think about consistency in the code: It would make sense to use one statement for all patch routes.

Reply:
Good points, so I think keeping your solution here makes the most sense. But as you said, it might be a good idea to make this a separate function that can be used in other patch routes as well (could include checking if a key actually exists and throwing an http exception if not). However, this can be done in the future and not as part of this PR.